File: Max-Points-on-a-Line/Max-Points-on-a-Line.cpp

/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
public:
    int maxPoints(vector<Point> &points) {
        unordered_map<float,int>mp;
        int ans=0;
        for(int i=0;i<points.size();i++)
        {
            mp.clear();
            mp[INT_MIN]=0;
            int cnt=1;
            for(int j=0;j<points.size();j++)
            {
                if(j==i)
                    continue;
                if(points[i].x==points[j].x && points[i].y==points[j].y)
                {
                    cnt++;
                    continue;
                }
                float k=points[i].x==points[j].x?INT_MAX:(float)(points[j].y-points[i].y)/(points[j].x-points[i].x);
                mp[k]++;
            }
            unordered_map<float,int>::iterator it=mp.begin();
            for( ;it!=mp.end();it++)
                ans=max(ans,it->second+cnt);
        }
        return ans;
    }
};


 

File: solutions/Max-Points-on-a-Line/Max-Points-on-a-Line.py


# Definition for a point
# class Point:
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution:
    # @param points, a list of Points
    # @return an integer
    def maxPoints(self, points):
        length=len(points)
        if length<3:
            return length
        ans=-1
        for i in range(length):
            slope={'int':0}
            cnt=1
            for j in range(length):
                if i==j:
                    continue;
                if points[i].x==points[j].x and points[i].y!=points[j].y:
                    slope['int']+=1
                elif points[i].x!=points[j].x:
                    k=1.0*(points[i].y-points[j].y)/(points[i].x-points[j].x)
                    slope[k]=1 if k not in slope else slope[k]+1
                else:
                    cnt+=1
            ans=max(ans,max(slope.values())+cnt)
        return ans
 